File: backend/model/alerts/alert_event.py
import datetime
import json
from dataclasses import dataclass

from model.alerts.alert_severity import AlertSeverity

#TODO: ADD ACKACTOR
@dataclass
class AlertEvent:
    def __init__(self, requires_ack: bool, severity: AlertSeverity, message: str, target_id: int, ack_time: datetime.datetime):
        self.alert_id = -1
        self.alert_time = datetime.datetime.now()
        self.ack_time = ack_time
        self.requires_ack = requires_ack
        self.severity = severity
        self.message = message
        self.target_id = target_id
        self.ws_notified = False
        self.db_notified = False
        self.acked = not requires_ack

    def to_dict(self) -> dict:
        alert_time = self.alert_time
        ack_time = self.ack_time
        if self.alert_time is not None:
            alert_time = alert_time.timestamp()

        if self.ack_time is not None:
            ack_time = self.ack_time.timestamp()

        return  {
            "alert-id": self.alert_id,
            "alert-time": alert_time,
            "ack-time": ack_time,
            "requires-ack": self.requires_ack,
            "severity": self.severity,
            "message": self.message,
            "target-id": self.target_id,
            "acked": self.acked,
        }

    def __str__(self):
        as_dict = self.to_dict()
        as_dict["severity"] = str(as_dict["severity"])
        return json.dumps(as_dict)

File: backend/src/alerts/alert_event.rs
use chrono::{DateTime, Utc};

use crate::{alerts::{AlertEvent, AlertSeverity}, types::{AlertRuleId, AlertTargetId}};


impl AlertEvent {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        requires_ack: bool,
        severity: AlertSeverity,
        message: String,
        target_id: AlertTargetId,
        ack_time: Option<DateTime<Utc>>,
        rule_id: AlertRuleId,
        value: String,
        ack_actor: Option<String>,
    ) -> Self {
        let now = Utc::now();
        AlertEvent {
            alert_id: -1,
            alert_time: Some(now),
            ack_time,
            requires_ack,
            severity,
            message,
            target_id,
            ws_notified: false,
            db_notified: false,
            acked: !requires_ack,
            rule_id: Some(rule_id),
            ack_actor,
            value,

        }
    }

    /// Convert to dictionary-like JSON value
    pub fn to_dict(&self, stringify: bool) -> serde_json::Value {
        let mut map = serde_json::Map::new();

        map.insert("alert-id".into(), serde_json::json!(self.alert_id));
        map.insert(
            "alert-time".into(),
            serde_json::json!(self.alert_time.map(|t| t.timestamp())),
        );
        map.insert(
            "ack-time".into(),
            serde_json::json!(self.ack_time.map(|t| t.timestamp())),
        );
        map.insert("requires-ack".into(), serde_json::json!(self.requires_ack));

        if stringify {
            map.insert("severity".into(), serde_json::json!(self.severity.to_string()));
        } else {
            map.insert("severity".into(), serde_json::to_value(self.severity).unwrap_or(serde_json::Value::String("".to_string())));
        }

        map.insert("message".into(), serde_json::json!(self.message));
        map.insert("target-id".into(), serde_json::json!(self.target_id));
        map.insert("acked".into(), serde_json::json!(self.acked));
        map.insert("value".into(), serde_json::json!(self.value));

        serde_json::Value::Object(map)
    }
}

impl std::fmt::Display for AlertEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut dict = self.to_dict(true);
        // force severity stringified
        if let Some(sev) = dict.get_mut("severity") {
            *sev = serde_json::json!(self.severity.to_string());
        }
        write!(f, "{}", dict)
    }
}
